web/modules/custom/dpl_logging/dpl_logging.module

```diff
@@ -0,0 +1,113 @@
+<?php
+
+use Drupal\Component\Utility\DiffArray;
+use Drupal\Core\StringTranslation\TranslatableMarkup;
+use Drupal\user\Entity\User;
+
+/**
+ * Implements hook_ENTITY_TYPE_presave().
+ *
+ * When updating a user, we want to log the changes done - this is part of
+ * a security policy by DDF.
+ * For most fields, the old and new value gets logged, but others may be
+ * ignored, or declared on a meta-level (such as password).
+ */
+function dpl_logging_user_presave(User $user): void {
+  $original_user = $user->original;
+
+  if (!($original_user instanceof User)) {
+    return;
+  }
+
+  $original_values = $original_user->toArray();
+  $values = $user->toArray();
+  $updates = DiffArray::diffAssocRecursive($values, $original_values);
+
+  // Some fields are so meta that we do not need to log their changes.
+  $ignored_updates = ['changed', 'metatag', 'field_password_expiration'];
+  $diffs = [];
+
+  foreach ($updates as $key => $value) {
+    if (in_array($key, $ignored_updates)) {
+      continue;
+    }
+
+    if ($key === 'roles') {
+      $original_roles = $original_user->getRoles();
+      $new_roles = $user->getRoles();
```

Comment:
Another more object-oriented and functional approach to achieving this could be in somewhat pseudo-code

```php
$fields = $user->getFields(include_computed: FALSE);
$ignored_fields = ['changed', 'metatag', 'field_password_expiration'];
$relevant_fields = array_filter($fields, ($field) => !in_array($field->getName(), $ignored_fields));
$updated_fields = array_filter($fields, ($field) => !$field->equals($original->get($field->getName()));
$diffs = array_map($updated_fields, ($field) => {
  return = t(
      '@label (@key): @original_value changed to @value',
      [
        '@label' => $label,
        '@key' => $field->getName(),
        '@original_value' => $original->get($field->getName())->toString(),
        '@value' => $field->toString(),
      ],
      ['context' => 'DPL logging']
    )->render();
})
```

To me we get a more structured approach to data handling by not ejecting from object to arrays so early. This also means that we cannot use `DiffArray` but as I see it `getFields()` and `equals()` on `FieldItemListInterface` achieves the same thing.

In the end it also means that we do not have to rely on `[0]['value'] ?? NULL` for data retrieval.

Reply:
I get your point, but generally speaking, I prefer my way, as I feel it's more readable

Reply:
I understand that the functional approach (array_filter, array_map etc.) requires some getting used to. You could use the object oriented features from Drupals classes combined with classic loops just as well.